src/features/device-provisioning/gracefully-deny-conflicting-registrations.ts

```diff
@@ -0,0 +1,69 @@
+import _ from 'lodash';
+import { sbvrUtils, errors, permissions } from '@balena/pinejs';
+import { API_KEY_EXISTS_CACHE_TIMEOUT } from '../../lib/config';
+import { createMultiLevelStore } from '../../infra/cache';
+import { checkDeviceExistsIsFrozen } from '../device-state/middleware';
+
+const { ConflictError } = errors;
+const { api } = sbvrUtils;
+
+const checkApiKeyExistsQuery = _.once(() =>
+	api.resin.prepare<{ key: string }>({
+		resource: 'api_key',
+		passthrough: { req: permissions.root },
+		id: {
+			key: { '@': 'key' },
+		},
+		options: {
+			$select: 'id',
+		},
+	}),
+);
+
+export const checkApiKeyExistsStore = createMultiLevelStore<boolean>(
+	'checkApiKeyExists',
+	{
+		default: {
+			// We only care to cache the api keys that already exist,
+			// so that we throw a conflict error earlier.
+			isCacheableValue: (value: any) => value === true,
+			ttl: API_KEY_EXISTS_CACHE_TIMEOUT,
+		},
+		local: false,
+	},
+	false,
+);
+
+const checkApiKeyExists = async (key: string) => {
+	return await checkApiKeyExistsStore.wrap(key, async () => {
```

Comment:
Any reason you're manually creating the store and using `.wrap` rather than using the memoize helper for it?

Reply:
@Page- b/c I'm using `isCacheableValue: (value: any) => value === true`, so that we do not cache keys that are not found, since in that case the cache would have to be invalidated right after. So instead of caching & deleting in the same register call, I thought we could avoid storing/caching it at the first place.

Reply:
I could switch it over to `multiCacheMemoizee` and throw to avoid caching in the cases that it doesn't make sense though if you prefer it.